### ISSUE TITLE
Be Able to Specify Path When Reporting Issue Through Reporter#report

### ISSUE THREAD

**Issue by philippfromme** [bpmn-io/bpmnlint#69]
When using [`Reporter#report`](https://github.com/bpmn-io/bpmnlint/blob/master/lib/test-rule.js#L11) one can only provide the element ID and the error message. There is no way to provide additional properties e.g. the path of the property that has caused the error to be reported. I'd image something along the lines of:

```javascript
Reporter.report('ServiceTask_1', '<zeebe:TaskDefinition> must have <zeebe:type> property', {
  path: [ 'extensionElements', 'values', 0, 'zeebe:type' ]
});
```

Adding an optional third argument wouldn't introduce any breaking changes.

---

Related to https://github.com/bpmn-io/internal-docs/issues/430

**Comment by nikku** [comment]
The fundamental BPMNLint feature we'Re talking about here is: Be able to address sub-paths within an element.

**Comment by nikku** [comment]
Another way to think about it: We could also turn this thing upside down.

No path is introduced, instead elements are reported using the standard way:

```
Reporter.report(zeebeTaskDefinition, 'must have <zeebe:type> property');

// or

Reporter.report(buildPath(serviceTask, zebeeTaskDefinition), 'must have <zeebe:type> property');
```

**Comment by philippfromme** [comment]
@nikku That's a good question. We could introduce this as a general feature, yes.

**Comment by philippfromme** [comment]
I will prototype a solution. Might not happen today though as I need to work on the lint rules themselves in order to demo.

**Comment by philippfromme** [comment]
The utility for getting the path would have to be available in the respective plugins (e.g. [bpmnlint-plugin-camunda-compat](https://github.com/camunda/bpmnlint-plugin-camunda-compat)). I'm not sure what `Reporter.report(buildPath(serviceTask, zebeeTaskDefinition), 'must have <zeebe:type> property');` would look like? `Reporter.report([ 'ServiceTask_1', 'extensionElements', 0 'type' ], 'must have <zeebe:type> property');`? That would introduce a breaking change and also require us to parse the path assuming the first key is the element's ID. I'm not a huge fan of that idea. My current approach looks like this:

```javascript
class Reporter {
  constructor({ moddleRoot, rule }) {
    this.rule = rule;
    this.moddleRoot = moddleRoot;
    this.messages = [];
    this.report = this.report.bind(this);
  }

  report(id, message, options) {
    let report = {
      id,
      message
    };

    if (options) {
      report = {
        ...report,
        options // optional
      };
    }

    this.messages.push(report);
  }
}
```

The rule itself would look something like this:

```javascript
// ...

return {
  message: 'Element of type <zeebe:TaskDefinition> must have <zeebe:type> property',
  path: [ ...getPath(taskDefinition, node), 'zeebe:type' ] // get the path of `taskDefinition` relative to `node`
};
```

As @smbea is going to start working on https://github.com/camunda/bpmnlint-plugin-camunda-compat/issues/7 we need to come to an agreement regarding the solution for this issue.

@nikku What are your thoughts? Shall we have a chat about this?

**Comment by nikku** [comment]
How would the CLI output look like in your idea?

**Comment by nikku** [comment]
We have to think this end-to-end, and if that means a major breaking change, I'd be fine with it.

**Comment by philippfromme** [comment]
I guess you'd expect something similar to:

```sh
> bpmnlint invoice.bpmn

/Projects/process-application/resources/invoice.bpmn
  ServiceTask_1#extensionElements.values.0    error    Element of type <zeebe:TaskDefinition> must have <zeebe:type> property    camunda-cloud-1-0

✖ 1 problem (1 error, 0 warnings)
```

In that case we'd probably make the path concept known to bpmnlint. 🤔

**Comment by nikku** [comment]
Had a short conversation with @philippfromme on the topic: 

* It makes sense to build path awareness into the core:
    * `Reporter#report(id, message, path: [string|number][])`
    * Report path in a sensible way via the CLI, too
* We cannot rely on `$parent` for traversal (bpmnlint, bpmn-js-properties-panel)
    * Our usage of `moddle` does not guarantee that it is being maintained properly
    * If we decide to rely on it, our usage shall be fail-safe

---

Session minutes: 

```javascript
// $parent set on import (moddle-xml)?

// $parent set during modeling
const foo = moddle.create('bpmn:ExtensionElements');

// XXX
const blub = { };
foo.values.push(blub); // blub = { $parent: foo }
foo.bar = { }; // bar = { $parent: foo }

foo.values.pop(); // blub = { }
blub.$parent = null;

foo.bar = null; // bar = { }

// philipp solution
// always use ModdleElement#get and ModdleElement#set
// get and set maintain $parent relationship ==> NOT HAPPENING AT THE MOMENT

// try to construct path on a best effort basis; make fail safe

reporter.report('foo', 'Sequence flow is missing condition', [ 'values', 0 ]);
reporter.report('foo', 'Sequence flow is missing condition', buildPath(fooElement, someChildElement));

const errors = [
  {
    id: 'foo',
    message: 'Sequecne flow...',
    path: [ 'values', 0 ] 
  }
];
```
